icu4j/main/tests/core/src/com/ibm/icu/dev/test/message2/FromJsonTest.java

```diff
@@ -0,0 +1,412 @@
+                .pattern("let $foo = {(bar)} {bar {$foo}}")
+                .expected("bar bar")
+                .build(),
+            new TestCase.Builder()
+                .pattern("let $foo = {(bar)} {bar {$foo}}")
+                .arguments(Args.of( "foo", "foo"))
+                // expectedJs: "bar foo"
+                // It is undefined if we allow arguments to override local variables, or it is an error.
+                // And undefined who wins if that happens, the local variable of the argument.
+                .expected("bar bar")
+                .build(),
+            new TestCase.Builder()
+                .pattern("let $foo = {$bar} {bar {$foo}}")
+                .arguments(Args.of( "bar", "foo"))
+                .expected("bar foo")
+                .build(),
+            new TestCase.Builder()
+                .pattern("let $foo = {$bar :number} {bar {$foo}}")
+                .arguments(Args.of( "bar", 4.2))
+                .expected("bar 4.2")
+                .build(),
+            new TestCase.Builder()
+                .pattern("let $foo = {$bar :number minimumFractionDigits=2} {bar {$foo}}")
+                .arguments(Args.of( "bar", 4.2))
+                .expected("bar 4.20")
+                .build(),
+            new TestCase.Builder().ignore("Maybe") // Because minimumFractionDigits=foo
+                .pattern("let $foo = {$bar :number minimumFractionDigits=foo} {bar {$foo}}")
+                .arguments(Args.of( "bar", 4.2))
+                .expected("bar 4.2")
+                .errors("invalid-type")
+                .build(),
+            new TestCase.Builder().ignore("Maybe. Function speciffic behavior.")
+                .pattern("let $foo = {$bar :number} {bar {$foo}}")
+                .arguments(Args.of( "bar", "foo"))
+                .expected("bar NaN")
+                .build(),
+            new TestCase.Builder()
+                .pattern("let $foo = {$bar} let $bar = {$baz} {bar {$foo}}")
+                .arguments(Args.of( "baz", "foo"))
+                // expectedJs: "bar foo"
+                // It is currently undefined if a local variable (like $foo)
+                // can reference a local variable that was not yet defined (like $bar).
+                // That is called hoisting and it is valid in JavaScript or Python.
+                // Not allowing that would prevent circular references.
+                .expected("bar {$bar}")
+                .build(),
+            new TestCase.Builder()
+                .patternJs("match {$foo} when (1) {one} when * {other}")
+                .pattern("match {$foo :select} when (1) {one} when * {other}")
+                .arguments(Args.of( "foo", "1"))
+                .expected("one")
+                .build(),
+            new TestCase.Builder()
+                .pattern("match {$foo :plural} when 1 {one} when * {other}")
+                .arguments(Args.of( "foo", "1")) // Should this be error? Plural on string?
```

Comment:
opinion: error -- we shouldn't go overboard trying to auto-convert input values, such as from string to number

Reply:
I agree with you.
That is why my implementation returns `"other"`

This Monday we discussed these differences, because I think they reflect either differences in opinions, or unclear spec. So we should address them.